google/resource_cloudbuild_build_trigger.go

```diff
@@ -0,0 +1,292 @@
+			},
+			"build": {
+				Type:        schema.TypeList,
+				Description: "Contents of the build template.",
+				Optional:    true,
+				ForceNew:    true,
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"images": &schema.Schema{
+							Type:     schema.TypeList,
+							Optional: true,
+							ForceNew: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"step": &schema.Schema{
+							Type:     schema.TypeList,
+							Optional: true,
+							ForceNew: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"name": &schema.Schema{
+										Type:     schema.TypeString,
+										Optional: true,
+										ForceNew: true,
+									},
+									"args": &schema.Schema{
+										Type:     schema.TypeString,
+										Optional: true,
+										ForceNew: true,
+									},
+								},
+							},
+						},
+						"tags": &schema.Schema{
+							Type:     schema.TypeList,
+							Optional: true,
+							ForceNew: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"trigger_template": &schema.Schema{
+				Optional: true,
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				ForceNew: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"branch_name": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+						"commit_sha": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+						"dir": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+						"project": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+						"repo_name": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+						"tag_name": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceCloudbuildBuildTriggerCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	project, err := getProject(d, config)
+	if err != nil {
+		return err
+	}
+
+	// Build the address parameter
+	buildTrigger := &cloudbuild.BuildTrigger{}
+
+	if v, ok := d.GetOk("description"); ok {
+		buildTrigger.Description = v.(string)
+	}
+
+	if d.Get("build.#").(int) > 0 {
+		build := &cloudbuild.Build{}
+		if v, ok := d.GetOk("build.0.images"); ok {
+			build.Images = convertStringArr(v.([]interface{}))
+		}
+		if v, ok := d.GetOk("build.0.tags"); ok {
+			build.Tags = convertStringArr(v.([]interface{}))
+		}
+		stepCount := d.Get("build.0.step.#").(int)
+		if stepCount > 0 {
```

Comment:
I'd convert this to use make unconditionally, and then `append()` rather than `build.Steps[s] = step`; I'm not going to insist on it.  Either way, we have a pattern for these complex data structures, we like to have them in their own functions, parallel to the flatteners you already wrote (we call them expanders).